bigquery_etl/util/common.py

```diff
@@ -162,6 +163,130 @@ def write_sql(output_dir, full_table_id, basename, sql, skip_existing=False):
         f.write("\n")
 
 
+def qualify_table_references_in_file(path: Path) -> str:
+    """Add project id and dataset id to table/view references and persistent udfs in a given query.
+
+    e.g.:
+    `table` -> `target_project.default_dataset.table`
+    `dataset.table` -> `target_project.dataset.table`
+
+    This allows a query to run in a different project than the sql dir it is located in
+    while referencing the same tables.
+    """
+    # sqlglot cannot handle scripts with variables and control statements
+    if re.search(
+        r"^\s*DECLARE\b", path.read_text(), flags=re.MULTILINE
+    ) or path.name in ("script.sql", "udf.sql"):
+        raise NotImplementedError(
+            "Cannot qualify table_references of query scripts or UDFs"
+        )
+
+    # determine the default target project and dataset from the path
+    target_project = Path(path).parent.parent.parent.name
+    default_dataset = Path(path).parent.parent.name
+
+    # sqlglot doesn't support Jinja, so we need to render the queries and
+    # init queries to raw SQL
+    sql_query = render(
+        path.name,
+        template_folder=path.parent,
+        format=False,
+        **DEFAULT_QUERY_TEMPLATE_VARS,
+    )
+    init_query = render(
+        path.name,
+        template_folder=path.parent,
+        format=False,
+        is_init=lambda: True,
+        metrics=MetricHub(),
+    )
+    # use sqlglot to get the SQL AST
+    init_query_statements = sqlglot.parse(
+        init_query,
+        read="bigquery",
+    )
+    sql_query_statements = sqlglot.parse(sql_query, read="bigquery")
+
+    # tuples of (table identifier, replacement string)
+    table_replacements: Set[Tuple[str, str]] = set()
+
+    # find all non-fully qualified table/view references including backticks
+    for query in [init_query_statements, sql_query_statements]:
+        for statement in query:
+            if statement is None:
+                continue
+
+            cte_names = {
+                cte.alias_or_name.lower() for cte in statement.find_all(sqlglot.exp.CTE)
+            }
+
+            table_aliases = {
+                cte.alias_or_name.lower()
+                for cte in statement.find_all(sqlglot.exp.TableAlias)
+            }
+
+            for table_expr in statement.find_all(sqlglot.exp.Table):
+                # existing table ref including backticks without alias
+                table_expr.set("alias", "")
+                reference_string = table_expr.sql(dialect="bigquery")
+
+                matched_cte = [
+                    re.match(
+                        rf"^{cte}(?![a-zA-Z0-9_])",
```

Comment:
Does a check for an optional `.` cover the same cases?  This would be a little clearer about what it's doing.  Also somewhat more robust since table names can have hyphens and accents even though we don't use them and the other regex in this function doesn't account for them
```suggestion
                        rf"^{cte}\.?",
```

Reply:
With the optional `.` there might be cases where the CTE is a substring of some table name, which would then match incorrectly

Reply:
Oh right it would need to be something like `^{cte}(\.?|$)`.  It's fine as-is though

Reply:
I think I'll leave it as is for now.